### PR TITLE
Added version increment to TrimExcess and EnsureCapacity in Dictionary

### LINKED ISSUE #31007: Dictionary version increment is missing for TrimExcess() and EnsureCapacity()
`Dictionary<K,V>` has a `version` field that is incremented by modification and tested by enumerators such that any modification of a dictionary during enumeration causes an exception. This is to flag that the enumeration may not lead to self consistent results.

Calls to TrimExcess() and EnsureCapacity() appear to not increment version even though they could modify the backing storage in such a way that they could break enumeration.

They should both have `version++`. 

https://github.com/dotnet/coreclr/pull/18854

### PR DESCRIPTION
**Pull request by davidkaya**
Fixes #31007.

### CODE REVIEW

**Comment by danmosemsft** [comment]
@davidkaya can you please add a test for each - that would fail without  your change.

**Comment by davidkaya** [comment]
@danmosemsft I've added tests. I've noticed something:
In `EnsureCapacity` on https://github.com/dotnet/corefx/pull/31062/files#diff-72146f1d0ccb2f246a39b3aaceef746dL973 the `currentCapacity (prime)` is compared with `capacity (non prime)`
and
in TrimExcess on https://github.com/dotnet/corefx/pull/31062/files#diff-72146f1d0ccb2f246a39b3aaceef746dL1010 the `currentCapacity (prime)` is compared with `newSize (prime - computed from non-prime capacity)`.

Shouldn't the behaviour be the same?

**Comment by stephentoub** [comment]
We should make the dictionary product changes with a coreclr PR rather than in corefx. CI here won't validate it.

**Comment by davidkaya** [comment]
I can port the PR to coreclr if necessary

**Comment by danmosemsft** [comment]
Oops, @stephentoub is correct. Please create one PR in Coreclr for the dictionary, and remove dictionary from this one. In the documentation it explains how to run local corefx tests against local coreclr product changes.

**Comment by danmosemsft** [comment]
@maryamariyan please see question about implementation correctness.

**Comment by davidkaya** [comment]
@danmosemsft @maryamariyan I've noticed that Dictionary tests contain `EnsureCapacity_Generic_RequestingLargerCapacity_DoesNotInvalidateEnumeration`. It seams that it was intended behaviour once. Should I remove those tests based on #31007 or based on this observation #31007 is invalid? There are no similar tests for `TrimExcess`. 

`Enumerator.MoveNext()` returns false if the `hashcode` of the next `Entry` is  `=< 0`. 
So basically if we call `EnsureCapacity` during enumeration nothing would change, because `Enumerator.MoveNext()` would return `false` on first `hashCode =< 0` occurrence. 
If we call `TrimExcess` during enumeration then again nothing would change because `TrimExcess` does not remove entries with `hashCode >= 0` thus would return false on first entry with `hashCode <= 0`.

**Comment by danmosemsft** [comment]
I can't look at the code right now but don't both of these by their nature potentially replace the buckets and entries arrays? The enumerator is going to continue with the old arrays. It seems like a bad state?

**Comment by Wraith2** [comment]
The enumerators keep an index but not a reference to the array so assigning to the array isn't a problem. 

Trim is currently unsafe because when the size of the array changes the entries are re-bucketed and re-hashed so you may no longer have an index to the same item (if it's an valid index to an item at all if you moved from a non-contiguous state to a contiguous one). This absolutely need a fix imo.

EnsureCapcity will re-bucket but not re-hash so the current index remains valid but you can end up with an enumerator that was unable to movenext but now can. Perhaps that's desirable? I think it could be confusing but since people seldom use enumerators directly and linq isn't likely to mutate the dictionary it can be left as it, i'd prefer it to invalidate.

**Comment by danmosemsft** [comment]
I'm inclined to say that we should add the `version++` to both rather than rationalize about the safety of the implementation (which could change later to be unsafe). If there is an important sceanrio for calling these during enumeration, we can revisit.

**Comment by davidkaya** [comment]
I will move the impl. part to coreclr today/tomorrow./

**Comment by davidkaya** [comment]
@danmosemsft I've moved the implementation to dotnet/coreclr#19096 and reverted the commit in this PR.

**Comment by danmosemsft** [comment]
We can rerun CI when we pick up next coreclr.

**Comment by danmosemsft** [comment]
In #31450
